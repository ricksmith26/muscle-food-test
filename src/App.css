
.App {
	text-align: center;
	display: flex;
    flex-direction: column;
    align-items: center;
	justify-content: center;
	align-items: center;
	align-content: center;
	
}

.App-logo {
    height: 40vmin;
    pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
    .App-logo {
        animation: App-logo-spin infinite 20s linear;
    }
}

.App-header {
    background-color: #282c34;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);
    color: white;
}

.App-link {
    color: #61dafb;
}

@keyframes App-logo-spin {
    from {
        transform: rotate(0deg);
    }

    to {
        transform: rotate(360deg);
    }
}

.centered {
	display: flex;
	flex-direction: row;
	flex-wrap: wrap;
	justify-content: center;
	align-items: center;
	align-content: center;
}

.products-container {
	height: 500px;
	width: 800px;
	border: 2px solid black;
	border-radius: 8px;
	position: relative;
	background-color: lightslategray;
	margin-top: 100px;
}

.product-box {
	height: 100px;
	width: 100px;
	border: 1px solid black;
	border-radius: 8px;
	display: flex;
	flex-direction: row;
	justify-content: center;
	align-items: center;
	align-content: center;
	background-color: lightblue;
}

.coin-item {
	height: 75px;
	width: 75px;
	border: 1px solid black;
	border-radius: 37.5px;
	display: flex;
	flex-direction: row;
	justify-content: center;
	align-items: center;
	align-content: center;
	cursor: pointer;
	background-color: yellow;
	margin: 12px;
}

.coin-total {
	height: 100px;
	width: 150px;
	border: 2px solid black;
	position: absolute;
	top: 8px;
	right: 16px;
	display: flex;
	flex-direction: row;
	justify-content: center;
	align-items: center;
	align-content: center;
	background-color: #FF8C00;
	border-radius: 8px;
}
.return-total {
	border: 2px solid black;
	position: absolute;
	top: 116px;
	right: 16px;
	display: flex;
	flex-direction: row;
	justify-content: center;
	align-items: center;
	align-content: center;
	background-color: #DC143C;
	border-radius: 8px;

}

.rejected-coins {
	height: 100px;
	width: 450px;
	border: 2px solid black;
	background-color: black;
	border-radius: 8px;
	color: white;
	position: absolute;
	bottom: 8px;
	display: flex;
	flex-direction: row;
	justify-content: space-around;
	align-items: center;
	align-content: center;
}

.coins {
	width: 500px;
	margin-top: 36px;
}