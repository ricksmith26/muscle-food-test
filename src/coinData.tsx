export const validCoins = {
    quarter: {
        weight: 5.67,
        diameter: 24.26,
        thickness: 1.75,
        value: 0.25
    },
    dime: {
        weight: 2.268,
        diameter: 17.91,
        thickness: 1.35,
        value: 0.10
    },
    nickel: {
        weight: 5,
        diameter: 21.21,
        thickness: 1.95,
        value: 0.05
    },
}

export const pocketChange = {
    dollar: {
        weight: 8.1,
        diameter: 30.61,
        thickness: 2,
        value: 100
    },
    quarter: {
        weight: 5.67,
        diameter: 24.26,
        thickness: 1.75,
        value: 0.25
    },
    dime: {
        weight: 2.268,
        diameter: 17.91,
        thickness: 1.35,
        value: 0.10
    },
    nickel: {
        weight: 5,
        diameter: 21.21,
        thickness: 1.95,
        value: 0.05
    },
    cent: {
        weight: 2.5,
        diameter: 19.05,
        thickness: 1.52,
        value: 0.01
    },
}

export const products = [
    {name: 'cola', cost: 1, qty: 1},
    {name: 'chips', cost: 0.50, qty: 10},
    {name: 'candy', cost: 0.65, qty: 10}
]
