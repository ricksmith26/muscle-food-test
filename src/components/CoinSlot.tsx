import React from 'react';

export default function CoinSlot(props: any) {

    return (
        <div>
            <div className="coin-total">{props.total === 0 && !props.error ? 'Insert Coins' : (props.error ? `${props.errorMSG}` : `Total: $${props.total.toFixed(2)}`)}</div>
            <button className="return-total" onClick={() => props.returnMoney()}>Return coins</button>
        </div>
    )
}