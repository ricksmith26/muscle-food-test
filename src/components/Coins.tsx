import React from 'react';

export default function Coins(props: any) {
    return (
        <div className="coins centered">

            {Object.entries(props.pocketChange).map((coin, i) => {
                return <button
                        className="coin-item"
                        onClick={() => props.countCoins(coin[1])} key={coin[0] + i}
                        id={coin[0]}>{coin[0]}</button>
            })}

        </div>
    )
}