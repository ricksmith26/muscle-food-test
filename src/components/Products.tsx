import React from 'react';
import { Product } from '../types';

export default function Products(props: any) {
    return (
        <div className="centered" style={{width: '600px'}}>
            {props.products.map((product: Product, i: number) => {
                return <button className="product-box" onClick={() => props.orderProduct(product, i)} key={product.name}>{`${product.name}: $${product.cost}`}</button>
            })}
        </div>
    )
}