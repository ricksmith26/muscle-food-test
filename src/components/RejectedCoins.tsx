import React from 'react';

export default function RejectedCoins(props: any) {
      
    return  (

            <div className="rejected-coins">

                <div data-testid="rejected-coins">Rejected Coins: {props.rejectedCoinsCount}</div>
                {props.changeGiven !== 0 ? <div data-testid="change-given">Change Given: ${props.changeGiven.toFixed(2)}</div> : null}

            </div>
            )
}