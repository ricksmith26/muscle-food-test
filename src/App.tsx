import React, {Component} from 'react';
import logo from './logo.svg';
import './App.css';

import {Coin, Product } from './types';
import {validCoins, pocketChange, products} from './coinData';
import Products from './components/Products';
import CoinSlot from './components/CoinSlot';
import RejectedCoins from './components/RejectedCoins'
import Coins from './components/Coins';

class App extends Component {

	state = {
		total: 0,
		rejectedCoinsCount: 0,
		change: 0,
		changeCoins: [],
		errorMSG: '',
		error: false,
		changeGiven: 0, 
		products: [],
		bank: 10
	}

	async componentDidMount() {
	}

	render() {
		return (
			<div className="App">
				Vending machine
				<div className="products-container centered">

					<Products orderProduct={this.orderProduct} products={products}></Products>

					<CoinSlot
						errors={this.state.error}
						returnMoney={this.returnMoney}
						total={this.state.total}
						></CoinSlot>
					
					<RejectedCoins
						rejectedCoinsCount={this.state.rejectedCoinsCount}
						changeGiven={this.state.changeGiven}
						countCoins={this.countCoins}
						></RejectedCoins>
					
				</div>

				<Coins
					pocketChange={pocketChange}
					countCoins={this.countCoins}
					></Coins>
				

			</div>
		);
	}
	
	countCoins = (addedCoin: Coin) => {
		const coinRefs = Object.values(validCoins)
		let validCoin = false;

		coinRefs.forEach((coinRef) => {
			if (coinRef.diameter === addedCoin.diameter
				&& coinRef.weight === addedCoin.weight
				&& coinRef.thickness === addedCoin.thickness) {
				validCoin = true
			}
		})

		if (validCoin) {
			this.setState({total: this.state.total + addedCoin.value }, () => console.log(this.state.total))
		} else {
			const rejectedCoinsCount = this.state.rejectedCoinsCount + 1;
			this.setState({rejectedCoinsCount})
		}
	}

	orderProduct = (product: Product, i: number) => {
		// if sold out
		if (product.qty === 0) {
			return this.setState({error: true, errorMSG: 'SOLD OUT'}
					, () =>{ 
					
					setTimeout(() => {
						this.setState({error: false})
					}, 5000)})
		}
		// checks enough money is given
		if (product.cost <= this.state.total) {
			products[i].qty-= 1; 
			this.setState({changeGiven: this.state.total - product.cost, total: 0, bank: this.state.bank + product.cost}, () => {
				setTimeout(() => {
					this.setState({changeGiven: 0})
				}, 5000)
			})
		}
		if (product.cost > this.state.total) {
			this.setState({error: true, errorMSG: `$${product.cost - this.state.total} needed`}, () => {
				this.setState({error: false, errorMSG: ''})
			})
		}
		// resets errors
		if (this.state.total === 0) {
			this.setState({total: 0, error: false, changeGiven: 0})
		}
		// error for exact change given created
		if (this.state.bank < 0.35) {
			this.setState({error: true, errorMSG: 'EXACT CHANGE NEEDED'})
			setTimeout(() => {
				this.setState({error: false, errorMSG: ''})
			}, 5000)
		}
	}

	returnMoney = () => {
		const change = this.state.total;
		this.setState({total: 0, changeGiven: change})
	}
}

export default App;
// import React from 'react';

// const App = () => <div className="app-container" >Vending machine</div>;

// export default App;