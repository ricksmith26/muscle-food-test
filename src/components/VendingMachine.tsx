import React, {Component} from 'react';

import {Coin, Product, VendingMachineState } from '../types';
import { products, validCoins } from '../coinData';
import Products from '../components/Products';
import CoinSlot from '../components/CoinSlot';
import RejectedCoins from '../components/RejectedCoins'
import Coins from '../components/Coins';

class VendingMachine extends Component {

	state = {
		total: 0,
		rejectedCoinsCount: 0,
		change: 0,
		changeGiven: 0, 
		bank: 0.15,
		errorMSG: '',
		error: false,
	}

	//adds coins to total
	public countCoins = (addedCoin: Coin) => {
		const coinRefs = Object.values(validCoins)
		let validCoin = false;

		coinRefs.forEach((coinRef) => {
			if (coinRef.diameter === addedCoin.diameter
				&& coinRef.weight === addedCoin.weight
				&& coinRef.thickness === addedCoin.thickness) {
				validCoin = true
			}
		})

		if (validCoin) {
			this.setState({total: this.state.total + addedCoin.value })
		} else {
			const rejectedCoinsCount = this.state.rejectedCoinsCount + 1;
			this.setState({rejectedCoinsCount})
		}
	}

	public orderProduct = (product: Product, i: number) => {

		// if sold out
		if (product.qty === 0) {
            return this.ifSoldOut();
		}
		// checks enough money is given
		if (product.cost <= this.state.total) {
			this.successfulPurchase(product, i)
		}
		//not Enough Funds
		if (product.cost > this.state.total) {
			this.notEnoughFunds(product)
		}
		// resets errors
		if (this.state.total === 0) {
			this.setState({total: 0, error: false, changeGiven: 0})
		}
		if (this.state.bank < 0.15) {
			this.notEnoughChangeInBank()
		}
    }
    
    public ifSoldOut() {
        return this.setState({error: true, errorMSG: 'SOLD OUT'}        
                , () =>{ 
                setTimeout(() => {
                    this.setState({error: false})
                }, 5000)})
    }

    public successfulPurchase(product: Product, i: number) {
        products[i].qty-= 1; 
        this.setState({changeGiven: this.state.total - product.cost, total: 0, bank: this.state.bank + product.cost}, () => {
                setTimeout(() => {
                    this.setState({changeGiven: 0})
                }, 5000)
            })
    }

    public notEnoughFunds(product: Product) {
        this.setState({error: true, errorMSG: `PRICE: $${product.cost}`},)
    }

    public notEnoughChangeInBank() {
        this.setState({error: true, errorMSG: 'EXACT CHANGE NEEDED'})
    }

	public returnMoney = () => {
		const change = this.state.total;
		this.setState({total: 0, changeGiven: change})
    }

    private getCoinSlotSearchBarProps = () => {

        return {
            error: this.state.error,
            errorMSG: this.state.errorMSG,
            returnMoney: this.returnMoney,
            total: this.state.total
        }
	}

	componentDidMount() {
		if (this.state.bank < 0.15) {
			this.setState({error: true, errorMSG: 'Exact Change'})
		}
	}
	
	componentDidUpdate(prevState: VendingMachineState) {
		if (prevState.bank !== this.state.bank && this.state.bank < 0.15) {
			this.setState({error: true, errorMSG: 'Exact Change'})
		}
	}

	render() {

		return ( 
			<div className="App">

				Vending machine

				<div className="products-container centered">

					<Products
						orderProduct={this.orderProduct}
						products={products}
						></Products>

					<CoinSlot
						{...this.getCoinSlotSearchBarProps()}
						></CoinSlot>
					
					<RejectedCoins
                        changeGiven={this.state.changeGiven}
                        rejectedCoinsCount={this.state.rejectedCoinsCount}
						></RejectedCoins>
					
				</div>

				<Coins countCoins={this.countCoins}></Coins>

			</div>
		);
	}

}

export default VendingMachine;