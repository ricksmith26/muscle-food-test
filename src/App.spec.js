import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import Products from './components/Products';
import CoinSlot from './components/CoinSlot';
import RejectedCoins from './components/RejectedCoins';
import VendingMachine from './components/VendingMachine';
import Coins from './components/Coins';
import { pocketChange, products } from './coinData';
import "../setupTests"

describe('App', () => {
	let wrapper;

	beforeEach(() => wrapper = shallow(<App />));

	it('should render correctly', () => expect(wrapper).toMatchSnapshot());

	it('should render a <div />', () => {

		expect(wrapper.find('div').length).toEqual(0);
	});	
});

describe('Vending Machine' , () => {

	let wrapper;

	beforeEach(() => wrapper = shallow(<VendingMachine />));

	it('should render the Products Component', () => {
		expect(wrapper.containsMatchingElement(<Products />)).toEqual(true);
	});

	it('should render the Coin Slot Component', () => {
		expect(wrapper.containsMatchingElement(<CoinSlot />)).toEqual(true);
	});

	it('should render the Rejected Coins Component', () => {
		expect(wrapper.containsMatchingElement(<RejectedCoins/>)).toEqual(true);
	});

	it('should render the Display Component', () => {
		expect(wrapper.containsMatchingElement(
		  <CoinSlot total={wrapper.instance().state.total} />
		)).toEqual(true);
	});
})


describe('Products', () => {
	let wrapper;

	beforeEach(() => wrapper = shallow(<Products products={products}/>));

	it('should render correctly', () => expect(wrapper).toMatchSnapshot());

	it('should render a <div />', () => {

		expect(wrapper.find('div').length).toEqual(1);
	});

	it('should render a buttons', () => {

		expect(wrapper.find('button').length).toEqual(3);
	});
})

describe('Coin Slot', () => {
	let wrapper;

	beforeEach(() => wrapper = shallow(<CoinSlot total={1}/>));

	it('should render correctly', () => expect(wrapper).toMatchSnapshot());

	it('should render a <div />', () => {

		expect(wrapper.find('div').length).toEqual(2);
	});

	it('should render a button', () => {

		expect(wrapper.find('button').length).toEqual(1);
	});

	it('renders the value of total', () => {
		wrapper.setProps({ total: 1 });
		expect(wrapper.text()).toMatch('Total: $1.00');
	});

})

describe('Coins', () => {
	let wrapper;

	beforeEach(() => wrapper = shallow(<Coins pocketChange={pocketChange} />));

	it('should render correctly', () => expect(wrapper).toMatchSnapshot());

	it('should render a <div />', () => {

		expect(wrapper.find('div').length).toEqual(1);
	})

	it('should add coins', () => {

		const mockCallBack = jest.fn();

		const button = shallow((<Coins
								pocketChange={pocketChange}
								countCoins={mockCallBack}jest
								>Ok!</Coins>));

		button.find('#quarter').prop('onClick')();
		
		expect(mockCallBack.mock.calls.length).toEqual(1);
	})
})

describe('Rejected Coins', () => {
	let wrapper;

	beforeEach(() => wrapper = shallow(<RejectedCoins changeGiven={0}/>));

	it('should render correctly', () => expect(wrapper).toMatchSnapshot());

	it('should render a <div />', () => {

		expect(wrapper.find('div').length).toEqual(2);

	});

	it('renders the count of rejected coinsl', () => {

		wrapper.setProps({ rejectedCoinsCount: 1 });

		expect(wrapper.find('.rejected-coins').text()).toMatch('Rejected Coins: 1');
	});

	it('renders the value of total change given', () => {

		wrapper.setProps({ changeGiven: 1.25 });

		expect(wrapper.find('.rejected-coins').text()).toMatch('Change Given: $1.25');
	});
	
})


