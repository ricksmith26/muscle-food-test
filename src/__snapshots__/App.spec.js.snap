// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `ShallowWrapper {}`;

exports[`Coin Slot should render correctly 1`] = `ShallowWrapper {}`;

exports[`Coins should render correctly 1`] = `ShallowWrapper {}`;

exports[`Products should render correctly 1`] = `ShallowWrapper {}`;

exports[`Rejected Coins should render correctly 1`] = `ShallowWrapper {}`;
